docs/design_notes/lift.md

```diff
@@ -0,0 +1,338 @@
+For example, we might decide to share parameters if we want to vectorize add a batch dimension to a model. 
+
+At the same time we can write generic code that uses transformations without knowing exactly what kind of variables the submodules will create.
+Collections thus strike a balance between fine-grained control and generality.
+We also avoid brittle string matching code that loops over all variables and tries to split up collections in an ad-hoc way based on
+naming conventions like: does this variable name have a prefix "kernel". 
+If more fine-grained control is necessary a user can simply split up a set of variables over multiple collections that should be handled differently.
+
+
+### Transformation vs variable control
+
+
+Lifting behavior could be defined either at the transformation level or during variable definition.
+We use transformation level definitions of lifting behavior.
+The reason for this choice is that there are many different transformations with various behaviors.
+For example: `vmap` has broadcasted and vectorized arguments, while `scan` has scan, carry, and broadcast arguments.
+A variable would have to define its behavior for all these transformations otherwise a `Module` would not be compatible with
+these tranformations. Or alternatively, we would have to make default decisions for how transformations are handled.
+However, this could lead to silent bugs becuase the behavior might not actually be valid given the users intent. 
+
+The lift package also provides a general purpose `transform` which allows an arbitrary function to transforms a variable collection.
+This can be used to for example tie the weights in a tied autoencoder by transposing the weights.
+It is unclear whether a similair general purpose transform could be defined if lifting decisions were made at variable definition.
+
+
+### Linen
+
+The lifting module does not know about linen Modules. Instead it operates directly on instances of `flax.core.Scope`.
+A `Scope` contains the variables and PRNG sequences of a `Module`. Each module has a `Scope` instance linked to it as long
+as its bound to a parent or it was created using a `init` or `apply`.
+
+When a `Module` is transformed we use the `flax.core.lift` APIs to lift the scope and use `Module.clone()` to create a new `Module` instance with the lifted `Scope` bound to it.
+
+`flax.linen.transforms` exposes wrappers for the transformations in `flax.core.lift`. The core lifting APIs operate on functions while
+the linen wrappers can transform either a `Module` class or a `Module` method.
+
+Lifting is thus implemented independently from the linen API. This seperation of concern simplifies the implementation while also allowing other Module APIs that chooce a different syntatic sugar for defining Modules to reuse the very generic and flexible `Scope `+ lift APIs for handling state and RNGs. 
```

Comment:
Suggestion:

- Nit picking
- Typo fix ("separation", "syntactic", "choose")
- `Module`, `lift` and `scope` (of `flax.core`) for consistency

```suggestion
Thus, lifting is implemented independently from the Linen API. This separation of concern simplifies the implementation, while also allowing other `Module` APIs that choose a different syntactic sugar for defining `Module`s to reuse the very generic and flexible `scope `and `lift` APIs for handling state and RNGs. 
```